File: src/main/java/me/newyith/fortress/util/Batch.java
package me.newyith.fortress.util;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Set;

public abstract class Batch extends BaseUUID {
	protected static class Model {
		private BaseUUID.Model superModel = null;
		private final transient BatchData batchData;

		@JsonCreator
		public Model(@JsonProperty("superModel") BaseUUID.Model superModel) {
			this.superModel = superModel;

			//rebuild transient fields
			this.batchData = BatchDataStore.get(superModel.uuid);
		}
	}
	protected Model model = null;

	@JsonCreator
	public Batch(@JsonProperty("model") Model model) {
		super(model.superModel);
		this.model = model;
	}

	public Batch(AuthToken authToken, Set<Point> points) {
		super(); //sets super.model
		BatchDataStore.put(super.model.uuid, new BatchData(authToken, points));
		model = new Model(super.model);
	}

	//-----------------------------------------------------------------------

	public boolean authorizedBy(AuthToken otherAuthToken) {
		return model.batchData.getAuthToken().equals(otherAuthToken);
	}

	public AuthToken getAuthToken() {
		return model.batchData.getAuthToken();
	}

	public Set<Point> getPoints() {
		return model.batchData.getPoints();
	}

	public boolean contains(Point p) {
		return model.batchData.getPoints().contains(p);
	}

	public void destroy() {
		BatchDataStore.remove(super.model.uuid);
	}
}

File: src/main/java/me/newyith/util/batch/Batch.kt
package me.newyith.util.batch

import com.fasterxml.jackson.annotation.JsonProperty
import me.newyith.util.Point
import me.newyith.util.BaseUUID
import me.newyith.util.Log

//Note: BatchGroupUUID is the new AuthToken
abstract class Batch(
	@JsonProperty("groupUuid") groupUuid: BatchGroupUUID
) : BaseUUID() {
	constructor(groupUuid: BatchGroupUUID, points: Set<Point>)
		: this(groupUuid)
	{
		val data = BatchData(groupUuid, points)
		BatchDataStore.forGroup(groupUuid)[uuid] = data
	}

	val batchData: BatchData by lazy {
		val data: BatchData? = BatchDataStore.forGroup(groupUuid)[uuid]
		when (data) {
			null -> {
				Log.error("Failed to find BatchData for groupUuid + uuid pair " + groupUuid + " + " + uuid)
				BatchData(groupUuid, HashSet()) //create fake data to keep things from blowing up more than necessary
			}
			else -> data
		}
	}

	val points: Set<Point>
		get() = batchData.points

	//isInGroup() is the new authorizedBy()
	fun isInGroup(otherGroupUuid: BatchGroupUUID): Boolean {
		return batchData.groupUuid == otherGroupUuid
	}

	operator fun contains(p: Point): Boolean {
		return batchData.points.contains(p)
	}

	fun destroy() {
		BatchDataStore.forGroup(batchData.groupUuid).remove(uuid)
	}
}
























//copy of orig:

//abstract class Batch : BaseUUID {
//	private var model: Model? = null
//
//	val points: Set<Point>
//		get() = model!!.batchData.points
//
//	protected class Model @JsonCreator
//	constructor(@JsonProperty("uuid") uuid: UUID) : BaseUUID.Model(uuid) {
//		@Transient private val batchData: BatchData
//
//		init {
//
//			//rebuild transient fields
//			this.batchData = BatchDataStore.get(uuid)
//		}
//	}
//
//	@JsonCreator
//	constructor(@JsonProperty("model") model: Model) : super(model) {
//		this.model = model
//	}
//
//	constructor(authToken: AuthToken, points: Set<Point>) {
//		BatchDataStore.put(super.getUuid(), BatchData(authToken, points))
//		model = Model(super.getUuid())
//	}
//
//	//-----------------------------------------------------------------------
//
//	fun authorizedBy(otherAuthToken: AuthToken): Boolean {
//		return model!!.batchData.authToken == otherAuthToken
//	}
//
//	operator fun contains(p: Point): Boolean {
//		return model!!.batchData.points.contains(p)
//	}
//
//	fun destroy() {
//		BatchDataStore.remove(super.getUuid())
//	}
//}
